From fea3f133bac6125fb4a495bcb29bdd4207ab496a Mon Sep 17 00:00:00 2001
From: Daniel Stone <daniels@collabora.com>
Date: Wed, 16 Nov 2016 10:54:10 +0000
Subject: [PATCH 69/94] compositor-drm: Add blob_id member to drm_mode

For atomic modesetting support, the mode is identified by a blob
property ID, rather than being passed inline. Add a blob_id member to
drm_mode to handle this, including refactoring mode destruction into a
helper function.

Differential Revision: https://phabricator.freedesktop.org/D1504

Signed-off-by: Daniel Stone <daniels@collabora.com>
---
 src/compositor-drm.c | 20 ++++++++++++++++----
 1 file changed, 16 insertions(+), 4 deletions(-)

diff --git a/src/compositor-drm.c b/src/compositor-drm.c
index ae6cdfe..ed89859 100644
--- a/src/compositor-drm.c
+++ b/src/compositor-drm.c
@@ -250,6 +250,7 @@ struct drm_backend {
 struct drm_mode {
 	struct weston_mode base;
 	drmModeModeInfo mode_info;
+	uint32_t blob_id;
 };
 
 enum drm_fb_type {
@@ -3253,6 +3254,7 @@ drm_output_add_mode(struct drm_output *output, const drmModeModeInfo *info)
 
 	mode->base.refresh = refresh;
 	mode->mode_info = *info;
+	mode->blob_id = 0;
 
 	if (info->type & DRM_MODE_TYPE_PREFERRED)
 		mode->base.flags |= WL_OUTPUT_MODE_PREFERRED;
@@ -3262,6 +3264,18 @@ drm_output_add_mode(struct drm_output *output, const drmModeModeInfo *info)
 	return mode;
 }
 
+/**
+ * Destroys a mode, and removes it from the list.
+ */
+static void
+drm_output_destroy_mode(struct drm_backend *backend, struct drm_mode *mode)
+{
+	if (mode->blob_id)
+		drmModeDestroyPropertyBlob(backend->drm.fd, mode->blob_id);
+	wl_list_remove(&mode->base.link);
+	free(mode);
+}
+
 static int
 drm_subpixel_to_wayland(int drm_value)
 {
@@ -4125,10 +4139,8 @@ err_output:
 	weston_output_destroy(&output->base);
 err_free:
 	wl_list_for_each_safe(drm_mode, next, &output->base.mode_list,
-							base.link) {
-		wl_list_remove(&drm_mode->base.link);
-		free(drm_mode);
-	}
+			      base.link)
+		drm_output_destroy_mode(b, drm_mode);
 
 	free(output);
 	free(config.modeline);
-- 
1.9.1

